<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Penalized Linear Mixed Models For Correlated Data • plmmr</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Penalized Linear Mixed Models For Correlated Data">
<meta name="description" content="Fits penalized linear mixed models that correct for unobserved confounding factors. plmmr infers and corrects for the presence of unobserved confounding effects such as population stratification and environmental heterogeneity. It then fits a linear model via penalized maximum likelihood. Originally designed for the multivariate analysis of single nucleotide polymorphisms (SNPs) measured in a genome-wide association study (GWAS), plmmr eliminates the need for subpopulation-specific analyses and post-analysis p-value adjustments. Functions for the appropriate processing of PLINK files are also supplied. For examples, see the package homepage. &lt;https://pbreheny.github.io/plmmr/&gt;.">
<meta property="og:description" content="Fits penalized linear mixed models that correct for unobserved confounding factors. plmmr infers and corrects for the presence of unobserved confounding effects such as population stratification and environmental heterogeneity. It then fits a linear model via penalized maximum likelihood. Originally designed for the multivariate analysis of single nucleotide polymorphisms (SNPs) measured in a genome-wide association study (GWAS), plmmr eliminates the need for subpopulation-specific analyses and post-analysis p-value adjustments. Functions for the appropriate processing of PLINK files are also supplied. For examples, see the package homepage. &lt;https://pbreheny.github.io/plmmr/&gt;.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/getting-started.html">Start here - getting to know plmmr</a></li>
    <li><a class="dropdown-item" href="articles/matrix_data.html">If your data is a matrix or data frame</a></li>
    <li><a class="dropdown-item" href="articles/plink_files.html">If your data is in PLINK files</a></li>
    <li><a class="dropdown-item" href="articles/delim_files.html">If your data is in (a) delimited file(s)</a></li>
    <li><a class="dropdown-item" href="articles/notation.html">Notes on notation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pbreheny/plmmr"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9">
<!-- badges: end -->
<div class="section level2">
<h2 id="plmmr-">plmmr <img src="reference/figures/plmmr_hex_sticker.png" align="right" width="150"><a class="anchor" aria-label="anchor" href="#plmmr-"></a>
</h2>
<p>The <code>plmmr</code> (<strong>p</strong>enalized <strong>l</strong>inear <strong>m</strong>ixed <strong>m</strong>odels in <strong>R</strong>) package contains functions that fit penalized linear mixed models to correct for unobserved confounding effects.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the latest version of the package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pbreheny/plmmr"</span><span class="op">)</span></span></code></pre></div>
<p>For a description of the motivation of the functions in this package (along with examples) refer to the second module of <a href="https://pbreheny.github.io/adv-gwas-tutorial/index.html" class="external-link">this GWAS data tutorial</a></p>
</div>
<div class="section level2">
<h2 id="so-how-fast-is-plmmr-and-how-well-does-it-scale">So how fast is <code>plmmr</code>? And how well does it scale?<a class="anchor" aria-label="anchor" href="#so-how-fast-is-plmmr-and-how-well-does-it-scale"></a>
</h2>
<p>To illustrate these important questions, I created a separate <a href="https://github.com/tabpeter/demo_plmmr/tree/master" class="external-link">GitHub repository</a> that has all the scripts for a <code>plmmr</code> workflow using publicly-available genome-wide association (GWAS) data. The main takeaway: using GWAS data from a study with 1,400 samples and 800,000 SNPs, a full <code>plmmr</code> analysis will run in about half an hour using a single core on a laptop.</p>
<p>Three smaller datasets ship with <code>plmmr</code>, and tutorials walking through how to analyze these data sets are documented in the <a href="https://pbreheny.github.io/plmmr/" class="external-link">documentation site</a>. While these datasets are useful for didactic purposes, they are not large enough to really highlight the computational scalability of <code>plmmr</code> – this is what motivated the creation of the separate repository for a GWAS workflow.</p>
</div>
<div class="section level2">
<h2 id="note-on-branches">Note on branches<a class="anchor" aria-label="anchor" href="#note-on-branches"></a>
</h2>
<p>The branches of this repo are organized in the following way:</p>
<ul>
<li><p><code>master</code> is the main (or ‘head’) branch.</p></li>
<li><p><code>gh_pages</code> is where we are keeping all the documentation for <code>plmmr</code></p></li>
<li><p><code>gwas_scale</code> is an <strong>archived</strong> branch that contains the development version of the package I used to run my dissertation analysis. Will delete this eventually.</p></li>
</ul>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing plmmr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Tabitha K. Peter <br><small class="roles"> Author </small> <a href="https://orcid.org/0009-0005-2524-4751" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Anna C. Reisetter <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-8332-4585" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Patrick J. Breheny <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-0650-1119" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Yujing Lu <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/pbreheny/plmmr" class="external-link"><img src="https://img.shields.io/static/v1?label=GitHub&amp;message=3.2.0.0&amp;color=blue&amp;logo=github" alt="GitHub version"></a></li>
<li><a href="https://github.com/pbreheny/plmmr/actions" class="external-link"><img src="https://github.com/pbreheny/plmmr/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/pbreheny/plmmr?branch=master" class="external-link"><img src="https://codecov.io/gh/pbreheny/plmmr/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tabitha K. Peter, Anna C. Reisetter, Patrick J. Breheny, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
