<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Preprocess PLINK files using the bigsnpr package — process_plink • plmmr</title><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><link href="../deps/KaTex-0.16.10/katex.min.css" rel="stylesheet"><script src="../deps/KaTex-0.16.10/katex.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Preprocess PLINK files using the bigsnpr package — process_plink"><meta name="description" content="Preprocess PLINK files using the bigsnpr package"><meta property="og:description" content="Preprocess PLINK files using the bigsnpr package"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/getting-started.html">Start here - getting to know plmmr</a></li>
    <li><a class="dropdown-item" href="../articles/matrix_data.html">If your data is a matrix or data frame</a></li>
    <li><a class="dropdown-item" href="../articles/plink_files.html">If your data is in PLINK files</a></li>
    <li><a class="dropdown-item" href="../articles/delim_files.html">If your data is in (a) delimited file(s)</a></li>
    <li><a class="dropdown-item" href="../articles/notation.html">Notes on notation</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/pbreheny/plmmr"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Preprocess PLINK files using the <code>bigsnpr</code> package</h1>
      <small class="dont-index">Source: <a href="https://github.com/pbreheny/plmmr/blob/master/R/process_plink.R" class="external-link"><code>R/process_plink.R</code></a></small>
      <div class="d-none name"><code>process_plink.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Preprocess PLINK files using the <code>bigsnpr</code> package</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">process_plink</span><span class="op">(</span></span>
<span>  <span class="va">data_dir</span>,</span>
<span>  <span class="va">data_prefix</span>,</span>
<span>  rds_dir <span class="op">=</span> <span class="va">data_dir</span>,</span>
<span>  <span class="va">rds_prefix</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  impute <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  impute_method <span class="op">=</span> <span class="st">"mode"</span>,</span>
<span>  id_var <span class="op">=</span> <span class="st">"IID"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-dir">data_dir<a class="anchor" aria-label="anchor" href="#arg-data-dir"></a></dt>
<dd><p>The path to the bed/bim/fam data files, <em>without</em> a trailing "/" (e.g., use <code>data_dir = '~/my_dir'</code>, <strong>not</strong> <code>data_dir = '~/my_dir/'</code>)</p></dd>


<dt id="arg-data-prefix">data_prefix<a class="anchor" aria-label="anchor" href="#arg-data-prefix"></a></dt>
<dd><p>The prefix (as a character string) of the bed/fam data files (e.g., <code>data_prefix = 'mydata'</code>)</p></dd>


<dt id="arg-rds-dir">rds_dir<a class="anchor" aria-label="anchor" href="#arg-rds-dir"></a></dt>
<dd><p>The path to the directory in which you want to create the new '.rds' and '.bk' files. Defaults to <code>data_dir</code></p></dd>


<dt id="arg-rds-prefix">rds_prefix<a class="anchor" aria-label="anchor" href="#arg-rds-prefix"></a></dt>
<dd><p>String specifying the user's preferred filename for the to-be-created .rds file (will be create insie <code>rds_dir</code> folder)
Note: 'rds_prefix' cannot be the same as 'data_prefix'</p></dd>


<dt id="arg-logfile">logfile<a class="anchor" aria-label="anchor" href="#arg-logfile"></a></dt>
<dd><p>Optional: the name (character string) of the prefix of the logfile to be written in 'rds_dir'. Default to NULL (no log file written).
Note: if you supply a file path in this argument, it will error out with a "file not found" error. Only supply the string; e.g., if you want my_log.log, supply 'my_log', the my_log.log file will appear in rds_dir.</p></dd>


<dt id="arg-impute">impute<a class="anchor" aria-label="anchor" href="#arg-impute"></a></dt>
<dd><p>Logical: should data be imputed? Default to TRUE.</p></dd>


<dt id="arg-impute-method">impute_method<a class="anchor" aria-label="anchor" href="#arg-impute-method"></a></dt>
<dd><p>If 'impute' = TRUE, this argument will specify the kind of imputation desired. Options are:
* mode (default): Imputes the most frequent call. See <code><a href="https://privefl.github.io/bigsnpr/reference/snp_fastImputeSimple.html" class="external-link">bigsnpr::snp_fastImputeSimple()</a></code> for details.
* random: Imputes sampling according to allele frequencies.
* mean0: Imputes the rounded mean.
* mean2: Imputes the mean rounded to 2 decimal places.
* xgboost: Imputes using an algorithm based on local XGBoost models. See <code><a href="https://privefl.github.io/bigsnpr/reference/snp_fastImpute.html" class="external-link">bigsnpr::snp_fastImpute()</a></code> for details. Note: this can take several minutes, even for a relatively small data set.</p></dd>


<dt id="arg-id-var">id_var<a class="anchor" aria-label="anchor" href="#arg-id-var"></a></dt>
<dd><p>String specifying which column of the PLINK <code>.fam</code> file has the unique sample identifiers. Options are "IID" (default) and "FID"</p></dd>


<dt id="arg-parallel">parallel<a class="anchor" aria-label="anchor" href="#arg-parallel"></a></dt>
<dd><p>Logical: should the computations within this function be run in parallel? Defaults to TRUE. See <code><a href="count_cores.html">count_cores()</a></code> and <code><a href="https://rdrr.io/pkg/bigparallelr/man/assert_cores.html" class="external-link">?bigparallelr::assert_cores</a></code> for more details.
In particular, the user should be aware that too much parallelization can make computations <em>slower</em>.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>Logical: should messages to be printed to the console be silenced? Defaults to FALSE</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>Logical: if existing <code>.bk</code>/<code>.rds</code> files exist for the specified directory/prefix, should these be overwritten? Defaults to FALSE. Set to TRUE if you want to change the imputation method you're using, etc.
<strong>Note</strong>: If there are multiple <code>.rds</code> files with names that start with "std_prefix_...", <strong>this will error out</strong>.
To protect users from accidentally deleting files with saved results, only one <code>.rds</code> file can be removed with this option.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Optional: additional arguments to <code><a href="https://privefl.github.io/bigsnpr/reference/snp_fastImpute.html" class="external-link">bigsnpr::snp_fastImpute()</a></code> (relevant only if impute_method = "xgboost")</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The filepath to the '.rds' object created; see details for explanation.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Three files are created in the location specified by <code>rds_dir</code>:</p><ul><li><p>'rds_prefix.rds': This is a list with three items:
(1) <code>X</code>: the filebacked <code><a href="https://rdrr.io/pkg/bigmemory/man/big.matrix.html" class="external-link">bigmemory::big.matrix</a></code> object pointing to the imputed genotype data.
This matrix has type 'double', which is important for downstream operations in <code><a href="create_design.html">create_design()</a></code>
(2) <code>map</code>: a data.frame with the PLINK 'bim' data (i.e., the variant information)
(3) <code>fam</code>: a data.frame with the PLINK 'fam' data (i.e., the pedigree information)</p></li>
<li><p>'prefix.bk': This is the
backingfile that stores the numeric data of the genotype matrix</p></li>
<li><p>'rds_prefix.desc'" This is the description file, as needed by the</p></li>
</ul><p>Note that <code>process_plink()</code> need only be run once for a given set of PLINK
files; in subsequent data analysis/scripts, <code><a href="get_data.html">get_data()</a></code> will access the '.rds' file.</p>
<p>For an example, see vignette on processing PLINK files</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tabitha K. Peter, Anna C. Reisetter, Patrick J. Breheny, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

