<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>If your data is in PLINK files • plmmr</title>
<script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><link href="../deps/KaTex-0.16.10/katex.min.css" rel="stylesheet">
<script src="../deps/KaTex-0.16.10/katex.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="If your data is in PLINK files">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting-started.html">Start here - getting to know plmmr</a></li>
    <li><a class="dropdown-item" href="../articles/matrix_data.html">If your data is a matrix or data frame</a></li>
    <li><a class="dropdown-item" href="../articles/plink_files.html">If your data is in PLINK files</a></li>
    <li><a class="dropdown-item" href="../articles/delim_files.html">If your data is in (a) delimited file(s)</a></li>
    <li><a class="dropdown-item" href="../articles/notation.html">Notes on notation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pbreheny/plmmr"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>If your data is in PLINK files</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/pbreheny/plmmr/blob/master/vignettes/articles/plink_files.Rmd" class="external-link"><code>vignettes/articles/plink_files.Rmd</code></a></small>
      <div class="d-none name"><code>plink_files.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pbreheny.github.io/plmmr/">plmmr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: bigalgebra</span></span>
<span><span class="co">#&gt; Loading required package: bigmemory</span></span></code></pre></div>
<p>A widely-used format for storing data from genome-wide association
studies (GWAS) is the <a href="https://www.cog-genomics.org/plink/1.9/formats" class="external-link">PLINK file
formats</a>, which consists of a triplet of ‘.bed’, ‘.bed’, and ‘.fam’
files. The <code>plmmr</code> package is equipped to analyze data from
PLINK files. If you have data in this format, keep reading – if you want
to know more about what each of these files contains, see <a href="https://pbreheny.github.io/adv-gwas-tutorial/quality_control.html" class="external-link">this
other tutorial</a> or the <a href="https://www.cog-genomics.org/plink/1.9" class="external-link">PLINK documentation</a>.
If your data is in delimited files (e.g., <code>.txt</code>,
<code>.csv</code>, etc.), read the article on analyzing data from
delimited files at
<code>vignette("delim_files", package = "plmmr")</code>.</p>
<p>The <code>plmmr</code> package is designed to handle data so that
users can analyze large data sets. For this reason, data must be
preprocessed into a specific format. There are two steps to prepare for
analysis: (1) process the data and (2) create a design. Processing the
data means that we take the feature data and create a ‘.rds’ object that
contains your feature data in a format compatible with the
<code>bigmemory</code> <a href="https://cran.r-project.org/web/packages/bigmemory/index.html" class="external-link">package</a>.
Creating a design involves standardizing the input of the features,
outcome, and penalty factor into the modeling functions
<code><a href="../reference/plmm.html">plmm()</a></code> and <code><a href="../reference/cv_plmm.html">cv_plmm()</a></code>.</p>
<div class="section level2">
<h2 id="processing-plink-files">Processing PLINK files<a class="anchor" aria-label="anchor" href="#processing-plink-files"></a>
</h2>
<p>First, unzip your PLINK files if they are zipped. Our example data,
<code>penncath_lite</code> data that ships with <code>plmmr</code> is
zipped; if you are on MacOS or Linux, you can run this command to
unzip:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># using a temp dir -- change to fit your preference</span></span>
<span><span class="fu"><a href="../reference/unzip_example_data.html">unzip_example_data</a></span><span class="op">(</span>outdir <span class="op">=</span> <span class="va">temp_dir</span><span class="op">)</span></span>
<span><span class="co">#&gt; Unzipped files are saved in /tmp/RtmppPF3Iz</span></span></code></pre></div>
<p>For GWAS data, we have to tell <code>plmmr</code> how to combine
information across all three PLINK files (the <code>.bed</code>,
<code>.bim</code>, and <code>.fam</code> files). We do this with
<code><a href="../reference/process_plink.html">process_plink()</a></code>.</p>
<p>Here, we will create the files we want in a temporary directory just
for the sake of example. Users can specify the folder of their choice
for <code>rds_dir</code>, as shown below:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># temp_dir &lt;- tempdir() # using a temporary directory (if you didn't already create one above)</span></span>
<span><span class="va">plink_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_plink.html">process_plink</a></span><span class="op">(</span>data_dir <span class="op">=</span> <span class="va">temp_dir</span>, </span>
<span>                            data_prefix <span class="op">=</span> <span class="st">"penncath_lite"</span>,</span>
<span>                            rds_dir <span class="op">=</span> <span class="va">temp_dir</span>,</span>
<span>                            rds_prefix <span class="op">=</span> <span class="st">"imputed_penncath_lite"</span>,</span>
<span>                            <span class="co"># imputing the mode to address missing values</span></span>
<span>                            impute_method <span class="op">=</span> <span class="st">"mode"</span>,</span>
<span>                            <span class="co"># overwrite existing files in temp_dir</span></span>
<span>                            <span class="co"># (you can turn this feature off if you need to)</span></span>
<span>                            overwrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            <span class="co"># turning off parallelization - </span></span>
<span>                            <span class="co">#   leaving this on causes problems knitting this vignette</span></span>
<span>                            parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Preprocessing penncath_lite data:</span></span>
<span><span class="co">#&gt; Creating penncath_lite.rds</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; There are 1401 observations and 4367 genomic features in the specified data files, representing chromosomes 1 - 22 </span></span>
<span><span class="co">#&gt; There are a total of 3514 SNPs with missing values</span></span>
<span><span class="co">#&gt; Of these, 13 are missing in at least 50% of the samples</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Imputing the missing (genotype) values using mode method</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; process_plink() completed </span></span>
<span><span class="co">#&gt; Processed files now saved as /tmp/RtmppPF3Iz/imputed_penncath_lite.rds</span></span></code></pre></div>
<p>You’ll see a lot of messages printed to the console here … the result
of all this is the creation of 3 files:
<code>imputed_penncath_lite.rds</code> and
<code>imputed_penncath_lite.bk</code> contain the data. <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Optionally, you can add a &lt;code&gt;.log&lt;/code&gt; file via the
&lt;code&gt;logfile&lt;/code&gt; argument: a log file is a text file documenting the
steps that were just done.&lt;/p&gt;"><sup>1</sup></a> These will show up in
the folder where the PLINK data is. What is returned is a filepath. The
<code>.rds</code> object at this filepath contains the processed data,
which we will now use to create our design.</p>
<p>For didactic purposes, let’s examine what’s in
<code>imputed_penncath_lite.rds</code> using the <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code>
function (<strong>Note</strong> Don’t do this in your analysis - the
section below reads the data into memory. This is just for
illustration):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">plink_data</span><span class="op">)</span> <span class="co"># notice: this is a `processed_plink` object </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">pen</span><span class="op">)</span> <span class="co"># note: genotype data is *not* in memory</span></span>
<span><span class="co">#&gt; List of 5</span></span>
<span><span class="co">#&gt;  $ X  :Formal class 'big.matrix.descriptor' [package "bigmemory"] with 1 slot</span></span>
<span><span class="co">#&gt;   .. ..@ description:List of 13</span></span>
<span><span class="co">#&gt;   .. .. ..$ sharedType: chr "FileBacked"</span></span>
<span><span class="co">#&gt;   .. .. ..$ filename  : chr "imputed_penncath_lite.bk"</span></span>
<span><span class="co">#&gt;   .. .. ..$ dirname   : chr "/tmp/RtmppPF3Iz/"</span></span>
<span><span class="co">#&gt;   .. .. ..$ totalRows : int 1401</span></span>
<span><span class="co">#&gt;   .. .. ..$ totalCols : int 4367</span></span>
<span><span class="co">#&gt;   .. .. ..$ rowOffset : num [1:2] 0 1401</span></span>
<span><span class="co">#&gt;   .. .. ..$ colOffset : num [1:2] 0 4367</span></span>
<span><span class="co">#&gt;   .. .. ..$ nrow      : num 1401</span></span>
<span><span class="co">#&gt;   .. .. ..$ ncol      : num 4367</span></span>
<span><span class="co">#&gt;   .. .. ..$ rowNames  : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ colNames  : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ type      : chr "double"</span></span>
<span><span class="co">#&gt;   .. .. ..$ separated : logi FALSE</span></span>
<span><span class="co">#&gt;  $ map:'data.frame': 4367 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ chromosome  : int [1:4367] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ marker.ID   : chr [1:4367] "rs3107153" "rs2455124" "rs10915476" "rs4592237" ...</span></span>
<span><span class="co">#&gt;   ..$ genetic.dist: int [1:4367] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   ..$ physical.pos: int [1:4367] 2056735 3188505 4275291 4280630 4286036 4302161 4364564 4388885 4606471 4643688 ...</span></span>
<span><span class="co">#&gt;   ..$ allele1     : chr [1:4367] "C" "T" "T" "G" ...</span></span>
<span><span class="co">#&gt;   ..$ allele2     : chr [1:4367] "T" "C" "C" "A" ...</span></span>
<span><span class="co">#&gt;  $ fam:'data.frame': 1401 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ family.ID  : int [1:1401] 10002 10004 10005 10007 10008 10009 10010 10011 10012 10013 ...</span></span>
<span><span class="co">#&gt;   ..$ sample.ID  : int [1:1401] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..$ paternal.ID: int [1:1401] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   ..$ maternal.ID: int [1:1401] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   ..$ sex        : int [1:1401] 1 2 1 1 1 1 1 2 1 2 ...</span></span>
<span><span class="co">#&gt;   ..$ affection  : int [1:1401] 1 1 2 1 2 2 2 1 2 -9 ...</span></span>
<span><span class="co">#&gt;  $ n  : int 1401</span></span>
<span><span class="co">#&gt;  $ p  : int 4367</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr "processed_plink"</span></span>
<span><span class="co"># notice: no more missing values in X</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pen</span><span class="op">$</span><span class="va">genotypes</span><span class="op">[</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-a-design">Creating a design<a class="anchor" aria-label="anchor" href="#creating-a-design"></a>
</h2>
<p>Now we are ready to create a <code>plmm_design</code>, which is an
object with the pieces we need for our model: a design matrix <span class="math inline">\mathbf{X}</span>, an outcome vector <span class="math inline">\mathbf{y}</span>, and a the vector with the penalty
factor indicators (1 = feature will be penalized, 0 = feature will not
be penalized).</p>
<p>As a side note: in GWAS studies, it is typical to include some
non-genomic factors as unpenalized covariates as part of the model. For
instance, you may want to adjust for sex or age (as shown in the example
below) – these are factors that you want to ensure are always included
in the selected model. The <code>plmmr</code> package allows you to
include these additional unpenalized predictors via the ‘add_predictor’
and ‘predictor_id’ options, both of which are passed through
<code><a href="../reference/create_design.html">create_design()</a></code> to the internal function
<code><a href="../reference/create_design_filebacked.html">create_design_filebacked()</a></code>. An example with these options
is included in the <code><a href="../reference/create_design.html">create_design()</a></code> documentation.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get outcome data </span></span>
<span><span class="va">penncath_pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="../reference/find_example_data.html">find_example_data</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">'penncath_clinical.csv'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>FamID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">penncath_pheno</span><span class="op">$</span><span class="va">FamID</span><span class="op">)</span>,</span>
<span>                   CAD <span class="op">=</span> <span class="va">penncath_pheno</span><span class="op">$</span><span class="va">CAD</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare a data.frame of the predictors for which we want to adjust: </span></span>
<span><span class="va">other_predictors</span> <span class="op">&lt;-</span> <span class="va">penncath_pheno</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'FamID'</span>, <span class="st">'sex'</span>, <span class="st">'age'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">other_predictors</span><span class="op">$</span><span class="va">FamID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">other_predictors</span><span class="op">$</span><span class="va">FamID</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pen_design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_design.html">create_design</a></span><span class="op">(</span>data_file <span class="op">=</span> <span class="va">plink_data</span>,</span>
<span>                            feature_id <span class="op">=</span> <span class="st">"FID"</span>,</span>
<span>                            rds_dir <span class="op">=</span> <span class="va">temp_dir</span>,</span>
<span>                            new_file <span class="op">=</span> <span class="st">"std_penncath_lite"</span>,</span>
<span>                            add_outcome <span class="op">=</span> <span class="va">phen</span>,</span>
<span>                            outcome_id <span class="op">=</span> <span class="st">"FamID"</span>,</span>
<span>                            outcome_col <span class="op">=</span> <span class="st">"CAD"</span>,</span>
<span>                            add_predictor <span class="op">=</span> <span class="va">other_predictors</span>,</span>
<span>                            predictor_id <span class="op">=</span> <span class="st">'FamID'</span>,</span>
<span>                            logfile <span class="op">=</span> <span class="st">"design"</span>,</span>
<span>                            <span class="co"># again, overwrite if needed; use with caution</span></span>
<span>                            overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Aligning external data with the feature data by FamID </span></span>
<span><span class="co">#&gt; Adding predictors from external data.</span></span>
<span><span class="co">#&gt; Aligning IDs between fam and predictor files</span></span>
<span><span class="co">#&gt; Column-wise combining data sets</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |                                                                      |   1%  |                                                                              |=                                                                     |   1%  |                                                                              |=                                                                     |   2%  |                                                                              |==                                                                    |   2%  |                                                                              |==                                                                    |   3%  |                                                                              |==                                                                    |   4%  |                                                                              |===                                                                   |   4%  |                                                                              |===                                                                   |   5%  |                                                                              |====                                                                  |   5%  |                                                                              |====                                                                  |   6%  |                                                                              |=====                                                                 |   6%  |                                                                              |=====                                                                 |   7%  |                                                                              |=====                                                                 |   8%  |                                                                              |======                                                                |   8%  |                                                                              |======                                                                |   9%  |                                                                              |=======                                                               |   9%  |                                                                              |=======                                                               |  10%  |                                                                              |=======                                                               |  11%  |                                                                              |========                                                              |  11%  |                                                                              |========                                                              |  12%  |                                                                              |=========                                                             |  12%  |                                                                              |=========                                                             |  13%  |                                                                              |=========                                                             |  14%  |                                                                              |==========                                                            |  14%  |                                                                              |==========                                                            |  15%  |                                                                              |===========                                                           |  15%  |                                                                              |===========                                                           |  16%  |                                                                              |============                                                          |  16%  |                                                                              |============                                                          |  17%  |                                                                              |============                                                          |  18%  |                                                                              |=============                                                         |  18%  |                                                                              |=============                                                         |  19%  |                                                                              |==============                                                        |  19%  |                                                                              |==============                                                        |  20%  |                                                                              |==============                                                        |  21%  |                                                                              |===============                                                       |  21%  |                                                                              |===============                                                       |  22%  |                                                                              |================                                                      |  22%  |                                                                              |================                                                      |  23%  |                                                                              |================                                                      |  24%  |                                                                              |=================                                                     |  24%  |                                                                              |=================                                                     |  25%  |                                                                              |==================                                                    |  25%  |                                                                              |==================                                                    |  26%  |                                                                              |===================                                                   |  26%  |                                                                              |===================                                                   |  27%  |                                                                              |===================                                                   |  28%  |                                                                              |====================                                                  |  28%  |                                                                              |====================                                                  |  29%  |                                                                              |=====================                                                 |  29%  |                                                                              |=====================                                                 |  30%  |                                                                              |=====================                                                 |  31%  |                                                                              |======================                                                |  31%  |                                                                              |======================                                                |  32%  |                                                                              |=======================                                               |  32%  |                                                                              |=======================                                               |  33%  |                                                                              |=======================                                               |  34%  |                                                                              |========================                                              |  34%  |                                                                              |========================                                              |  35%  |                                                                              |=========================                                             |  35%  |                                                                              |=========================                                             |  36%  |                                                                              |==========================                                            |  36%  |                                                                              |==========================                                            |  37%  |                                                                              |==========================                                            |  38%  |                                                                              |===========================                                           |  38%  |                                                                              |===========================                                           |  39%  |                                                                              |============================                                          |  39%  |                                                                              |============================                                          |  40%  |                                                                              |============================                                          |  41%  |                                                                              |=============================                                         |  41%  |                                                                              |=============================                                         |  42%  |                                                                              |==============================                                        |  42%  |                                                                              |==============================                                        |  43%  |                                                                              |==============================                                        |  44%  |                                                                              |===============================                                       |  44%  |                                                                              |===============================                                       |  45%  |                                                                              |================================                                      |  45%  |                                                                              |================================                                      |  46%  |                                                                              |=================================                                     |  46%  |                                                                              |=================================                                     |  47%  |                                                                              |=================================                                     |  48%  |                                                                              |==================================                                    |  48%  |                                                                              |==================================                                    |  49%  |                                                                              |===================================                                   |  49%  |                                                                              |===================================                                   |  50%  |                                                                              |===================================                                   |  51%  |                                                                              |====================================                                  |  51%  |                                                                              |====================================                                  |  52%  |                                                                              |=====================================                                 |  52%  |                                                                              |=====================================                                 |  53%  |                                                                              |=====================================                                 |  54%  |                                                                              |======================================                                |  54%  |                                                                              |======================================                                |  55%  |                                                                              |=======================================                               |  55%  |                                                                              |=======================================                               |  56%  |                                                                              |========================================                              |  56%  |                                                                              |========================================                              |  57%  |                                                                              |========================================                              |  58%  |                                                                              |=========================================                             |  58%  |                                                                              |=========================================                             |  59%  |                                                                              |==========================================                            |  59%  |                                                                              |==========================================                            |  60%  |                                                                              |==========================================                            |  61%  |                                                                              |===========================================                           |  61%  |                                                                              |===========================================                           |  62%  |                                                                              |============================================                          |  62%  |                                                                              |============================================                          |  63%  |                                                                              |============================================                          |  64%  |                                                                              |=============================================                         |  64%  |                                                                              |=============================================                         |  65%  |                                                                              |==============================================                        |  65%  |                                                                              |==============================================                        |  66%  |                                                                              |===============================================                       |  66%  |                                                                              |===============================================                       |  67%  |                                                                              |===============================================                       |  68%  |                                                                              |================================================                      |  68%  |                                                                              |================================================                      |  69%  |                                                                              |=================================================                     |  69%  |                                                                              |=================================================                     |  70%  |                                                                              |=================================================                     |  71%  |                                                                              |==================================================                    |  71%  |                                                                              |==================================================                    |  72%  |                                                                              |===================================================                   |  72%  |                                                                              |===================================================                   |  73%  |                                                                              |===================================================                   |  74%  |                                                                              |====================================================                  |  74%  |                                                                              |====================================================                  |  75%  |                                                                              |=====================================================                 |  75%  |                                                                              |=====================================================                 |  76%  |                                                                              |======================================================                |  76%  |                                                                              |======================================================                |  77%  |                                                                              |======================================================                |  78%  |                                                                              |=======================================================               |  78%  |                                                                              |=======================================================               |  79%  |                                                                              |========================================================              |  79%  |                                                                              |========================================================              |  80%  |                                                                              |========================================================              |  81%  |                                                                              |=========================================================             |  81%  |                                                                              |=========================================================             |  82%  |                                                                              |==========================================================            |  82%  |                                                                              |==========================================================            |  83%  |                                                                              |==========================================================            |  84%  |                                                                              |===========================================================           |  84%  |                                                                              |===========================================================           |  85%  |                                                                              |============================================================          |  85%  |                                                                              |============================================================          |  86%  |                                                                              |=============================================================         |  86%  |                                                                              |=============================================================         |  87%  |                                                                              |=============================================================         |  88%  |                                                                              |==============================================================        |  88%  |                                                                              |==============================================================        |  89%  |                                                                              |===============================================================       |  89%  |                                                                              |===============================================================       |  90%  |                                                                              |===============================================================       |  91%  |                                                                              |================================================================      |  91%  |                                                                              |================================================================      |  92%  |                                                                              |=================================================================     |  92%  |                                                                              |=================================================================     |  93%  |                                                                              |=================================================================     |  94%  |                                                                              |==================================================================    |  94%  |                                                                              |==================================================================    |  95%  |                                                                              |===================================================================   |  95%  |                                                                              |===================================================================   |  96%  |                                                                              |====================================================================  |  96%  |                                                                              |====================================================================  |  97%  |                                                                              |====================================================================  |  98%  |                                                                              |===================================================================== |  98%  |                                                                              |===================================================================== |  99%  |                                                                              |======================================================================|  99%  |                                                                              |======================================================================| 100%</span></span>
<span><span class="co">#&gt; There are 62 constant features in the data</span></span>
<span><span class="co">#&gt; Subsetting data to exclude constant features (e.g., monomorphic SNPs)</span></span>
<span><span class="co">#&gt; Column-standardizing the design matrix...</span></span>
<span><span class="co">#&gt; Standardization completed at 2025-02-18 18:02:46</span></span>
<span><span class="co">#&gt; Done with standardization. File formatting in progress</span></span>
<span></span>
<span><span class="co"># examine the design - notice the components of this object </span></span>
<span><span class="va">pen_design_rds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">pen_design</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">pen_design_rds</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 18</span></span>
<span><span class="co">#&gt;  $ X_colnames    : chr [1:4367] "rs3107153" "rs2455124" "rs10915476" "rs4592237" ...</span></span>
<span><span class="co">#&gt;  $ X_rownames    : chr [1:1401] "10002" "10004" "10005" "10007" ...</span></span>
<span><span class="co">#&gt;  $ n             : int 1401</span></span>
<span><span class="co">#&gt;  $ p             : int 4367</span></span>
<span><span class="co">#&gt;  $ is_plink      : logi TRUE</span></span>
<span><span class="co">#&gt;  $ outcome_idx   : int [1:1401] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;  $ y             : Named int [1:1401] 1 1 1 1 1 1 1 1 1 0 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:1401] "CAD1" "CAD2" "CAD3" "CAD4" ...</span></span>
<span><span class="co">#&gt;  $ std_X_rownames: chr [1:1401] "10002" "10004" "10005" "10007" ...</span></span>
<span><span class="co">#&gt;  $ unpen         : int [1:2] 1 2</span></span>
<span><span class="co">#&gt;  $ unpen_colnames: chr [1:2] "sex" "age"</span></span>
<span><span class="co">#&gt;  $ ns            : int [1:4307] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;  $ std_X_colnames: chr [1:4307] "sex" "age" "rs3107153" "rs2455124" ...</span></span>
<span><span class="co">#&gt;  $ std_X         :Formal class 'big.matrix.descriptor' [package "bigmemory"] with 1 slot</span></span>
<span><span class="co">#&gt;   .. ..@ description:List of 13</span></span>
<span><span class="co">#&gt;   .. .. ..$ sharedType: chr "FileBacked"</span></span>
<span><span class="co">#&gt;   .. .. ..$ filename  : chr "std_penncath_lite.bk"</span></span>
<span><span class="co">#&gt;   .. .. ..$ dirname   : chr "/tmp/RtmppPF3Iz/"</span></span>
<span><span class="co">#&gt;   .. .. ..$ totalRows : int 1401</span></span>
<span><span class="co">#&gt;   .. .. ..$ totalCols : int 4307</span></span>
<span><span class="co">#&gt;   .. .. ..$ rowOffset : num [1:2] 0 1401</span></span>
<span><span class="co">#&gt;   .. .. ..$ colOffset : num [1:2] 0 4307</span></span>
<span><span class="co">#&gt;   .. .. ..$ nrow      : num 1401</span></span>
<span><span class="co">#&gt;   .. .. ..$ ncol      : num 4307</span></span>
<span><span class="co">#&gt;   .. .. ..$ rowNames  : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ colNames  : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ type      : chr "double"</span></span>
<span><span class="co">#&gt;   .. .. ..$ separated : logi FALSE</span></span>
<span><span class="co">#&gt;  $ std_X_n       : num 1401</span></span>
<span><span class="co">#&gt;  $ std_X_p       : num 4307</span></span>
<span><span class="co">#&gt;  $ std_X_center  : num [1:4307] 1.33119 55.72448 0.00785 0.35974 1.01213 ...</span></span>
<span><span class="co">#&gt;  $ std_X_scale   : num [1:4307] 0.4706 9.3341 0.0883 0.7783 0.8636 ...</span></span>
<span><span class="co">#&gt;  $ penalty_factor: num [1:4307] 0 0 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr "plmm_design"</span></span></code></pre></div>
<p>A key part of what <code><a href="../reference/create_design.html">create_design()</a></code> is doing is
standardizing the columns of the genotype matrix. Below is a didactic
example showing that the columns of the <code>std_X</code> element in
our design have mean = 0 and variance = 1. <strong>Note again</strong>
that this is not something you should do in your analysis – this reads
the data into memory.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we can check to see that our data have been standardized </span></span>
<span><span class="va">std_X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bigmemory/man/attach.big.matrix.html" class="external-link">attach.big.matrix</a></span><span class="op">(</span><span class="va">pen_design_rds</span><span class="op">$</span><span class="va">std_X</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">std_X</span><span class="op">[</span>,<span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># columns have mean zero...</span></span>
<span><span class="co">#&gt;       Min.    1st Qu.     Median       Mean    3rd Qu.       Max. </span></span>
<span><span class="co">#&gt; -1.356e-16 -2.336e-17  3.814e-19  1.006e-18  2.524e-17  2.635e-16</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">std_X</span><span class="op">[</span>,<span class="op">]</span>, <span class="fl">2</span>, <span class="va">var</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># ... &amp; variance 1</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   1.001   1.001   1.001   1.001   1.001   1.001</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fitting-a-model">Fitting a model<a class="anchor" aria-label="anchor" href="#fitting-a-model"></a>
</h2>
<p>Now that we have a design object, we are ready to fit a model. By
default, the model fitting results are saved as files in the folder
specified in the <code>rds_dir</code> argument of <code>plmmm</code>. If
you want to return the model fitting results, set
<code>return_fit = TRUE</code> in <code><a href="../reference/plmm.html">plmm()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pen_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plmm.html">plmm</a></span><span class="op">(</span>design <span class="op">=</span> <span class="va">pen_design</span>,</span>
<span>                trace <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                return_fit <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; Note: The design matrix is being returned as a file-backed big.matrix object -- see bigmemory::big.matrix() documentation for details.</span></span>
<span><span class="co">#&gt; Reminder: the X that is returned here is column-standardized</span></span>
<span><span class="co">#&gt; Warning in plmm_checks(design, K = K, diag_K = diag_K, eta_star = eta_star, : Note: plmmr depends on the package bigalgebra, the current GitHub </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             version of which is throwing some warnings for filebacked analysis. </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             See https://github.com/fbertran/bigalgebra/issues/2 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             If you see a warning about 'stack imbalance' appear while you are </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             fitting a model with plmm() or cv_plmm(), we recommend downloading </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             the last stable version of bigalebra.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             This message is displayed after your first call to fit a model </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             using filebacked data.</span></span>
<span><span class="co">#&gt; Input data passed all checks at  2025-02-18 18:02:47</span></span>
<span><span class="co">#&gt; Starting decomposition.</span></span>
<span><span class="co">#&gt; Calculating the eigendecomposition of K</span></span>
<span><span class="co">#&gt; Eigendecomposition finished at  2025-02-18 18:02:49</span></span>
<span><span class="co">#&gt; Beginning rotation ('preconditioning').</span></span>
<span><span class="co">#&gt; Rotation (preconditiong) finished at  2025-02-18 18:02:49</span></span>
<span><span class="co">#&gt; Setting up lambda/preparing for model fitting.</span></span>
<span><span class="co">#&gt; Beginning model fitting.</span></span>
<span><span class="co">#&gt; Model fitting finished at  2025-02-18 18:02:52 </span></span>
<span><span class="co">#&gt; Beta values are estimated -- almost done!</span></span>
<span><span class="co">#&gt; Formatting results (backtransforming coefs. to original scale).</span></span>
<span><span class="co">#&gt; Model ready at  2025-02-18 18:02:52</span></span>
<span><span class="co"># you can turn off the trace messages by letting trace = F (default)</span></span></code></pre></div>
<p>We examine our model results below:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">pen_fit</span>, idx <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">#&gt; lasso-penalized regression model with n=1401, p=4370 at lambda=0.00964</span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt; The model converged </span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt; # of non-zero coefficients:  650 </span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pen_fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="plink_files_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="cross-validation">Cross validation<a class="anchor" aria-label="anchor" href="#cross-validation"></a>
</h2>
<p>To choose a tuning parameter for a model, <code>plmmr</code> offers a
cross validation method:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cv_plmm.html">cv_plmm</a></span><span class="op">(</span>design <span class="op">=</span> <span class="va">pen_design</span>,</span>
<span>                  type <span class="op">=</span> <span class="st">"blup"</span>,</span>
<span>                  return_fit <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                  trace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; Note: The design matrix is being returned as a file-backed big.matrix object -- see bigmemory::big.matrix() documentation for details.</span></span>
<span><span class="co">#&gt; Reminder: the X that is returned here is column-standardized</span></span>
<span><span class="co">#&gt; Starting decomposition.</span></span>
<span><span class="co">#&gt; Calculating the eigendecomposition of K</span></span>
<span><span class="co">#&gt; Beginning rotation ('preconditioning').</span></span>
<span><span class="co">#&gt; Rotation (preconditiong) finished at  2025-02-18 18:02:54</span></span>
<span><span class="co">#&gt; Setting up lambda/preparing for model fitting.</span></span>
<span><span class="co">#&gt; Beginning model fitting.</span></span>
<span><span class="co">#&gt; Model fitting finished at  2025-02-18 18:02:57 </span></span>
<span><span class="co">#&gt; 'Fold' argument is either NULL or missing; assigning folds randomly (by default).</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt; To specify folds for each observation, supply a vector with fold assignments.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Starting cross validation</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%Beginning eigendecomposition in fold  1 :</span></span>
<span><span class="co">#&gt; Starting decomposition.</span></span>
<span><span class="co">#&gt; Calculating the eigendecomposition of K</span></span>
<span><span class="co">#&gt; Fitting model in fold  1 :</span></span>
<span><span class="co">#&gt; Beginning rotation ('preconditioning').</span></span>
<span><span class="co">#&gt; Rotation (preconditiong) finished at  2025-02-18 18:02:58</span></span>
<span><span class="co">#&gt; Beginning model fitting.</span></span>
<span><span class="co">#&gt; Model fitting finished at  2025-02-18 18:03:00 </span></span>
<span><span class="co">#&gt;   |                                                                              |==============                                                        |  20%</span></span>
<span><span class="co">#&gt; Beginning eigendecomposition in fold  2 :</span></span>
<span><span class="co">#&gt; Starting decomposition.</span></span>
<span><span class="co">#&gt; Calculating the eigendecomposition of K</span></span>
<span><span class="co">#&gt; Fitting model in fold  2 :</span></span>
<span><span class="co">#&gt; Beginning rotation ('preconditioning').</span></span>
<span><span class="co">#&gt; Rotation (preconditiong) finished at  2025-02-18 18:03:01</span></span>
<span><span class="co">#&gt; Beginning model fitting.</span></span>
<span><span class="co">#&gt; Model fitting finished at  2025-02-18 18:03:03 </span></span>
<span><span class="co">#&gt;   |                                                                              |============================                                          |  40%Beginning eigendecomposition in fold  3 :</span></span>
<span><span class="co">#&gt; Starting decomposition.</span></span>
<span><span class="co">#&gt; Calculating the eigendecomposition of K</span></span>
<span><span class="co">#&gt; Fitting model in fold  3 :</span></span>
<span><span class="co">#&gt; Beginning rotation ('preconditioning').</span></span>
<span><span class="co">#&gt; Rotation (preconditiong) finished at  2025-02-18 18:03:04</span></span>
<span><span class="co">#&gt; Beginning model fitting.</span></span>
<span><span class="co">#&gt; Model fitting finished at  2025-02-18 18:03:07 </span></span>
<span><span class="co">#&gt;   |                                                                              |==========================================                            |  60%Beginning eigendecomposition in fold  4 :</span></span>
<span><span class="co">#&gt; Starting decomposition.</span></span>
<span><span class="co">#&gt; Calculating the eigendecomposition of K</span></span>
<span><span class="co">#&gt; Fitting model in fold  4 :</span></span>
<span><span class="co">#&gt; Beginning rotation ('preconditioning').</span></span>
<span><span class="co">#&gt; Rotation (preconditiong) finished at  2025-02-18 18:03:08</span></span>
<span><span class="co">#&gt; Beginning model fitting.</span></span>
<span><span class="co">#&gt; Model fitting finished at  2025-02-18 18:03:10 </span></span>
<span><span class="co">#&gt;   |                                                                              |========================================================              |  80%Beginning eigendecomposition in fold  5 :</span></span>
<span><span class="co">#&gt; Starting decomposition.</span></span>
<span><span class="co">#&gt; Calculating the eigendecomposition of K</span></span>
<span><span class="co">#&gt; Fitting model in fold  5 :</span></span>
<span><span class="co">#&gt; Beginning rotation ('preconditioning').</span></span>
<span><span class="co">#&gt; Rotation (preconditiong) finished at  2025-02-18 18:03:11</span></span>
<span><span class="co">#&gt; Beginning model fitting.</span></span>
<span><span class="co">#&gt; Model fitting finished at  2025-02-18 18:03:14 </span></span>
<span><span class="co">#&gt;   |                                                                              |======================================================================| 100%</span></span></code></pre></div>
<p>There are plot and summary methods for CV models as well:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cv_fit</span><span class="op">)</span> <span class="co"># summary at lambda value that minimizes CV error</span></span>
<span><span class="co">#&gt; lasso-penalized model with n=1401 and p=4370</span></span>
<span><span class="co">#&gt; At minimum cross-validation error (lambda=0.0424):</span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt;   Nonzero coefficients: 2</span></span>
<span><span class="co">#&gt;   Cross-validation error (deviance): 0.18</span></span>
<span><span class="co">#&gt;   Scale estimate (sigma): 0.418</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cv_fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="plink_files_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="details-create_design-for-plink-data">Details: <code>create_design()</code> for PLINK data<a class="anchor" aria-label="anchor" href="#details-create_design-for-plink-data"></a>
</h2>
<p>The call to <code><a href="../reference/create_design.html">create_design()</a></code> involves these steps:</p>
<ul>
<li><p>Integrate in the external phenotype information, if supplied.
<strong>Note: Any samples in the PLINK data that do not have a phenotype
value in the specified additional phenotype file will be removed from
the analysis.</strong></p></li>
<li><p>Identify missing values in both samples and
SNPs/features.</p></li>
<li><p>Impute missing values per user’s specified method. See <a href="https://privefl.github.io/bigsnpr/reference/snp_fastImputeSimple.html" class="external-link">R
documentation for <code>bigsnpr::snp_fastImputeSimple()</code></a> for
more details. <strong>Note: the plmmr package cannot fit models if
datasets have missing values.</strong> All missing values must be
imputed or subset out before analysis.</p></li>
<li><p>Integrate in the external predictor information, if supplied.
This could be a matrix of meta-data (e.g., age, principal components,
etc.). <strong>Note:</strong> If there are samples in the supplied file
that are not included in the PLINK data, these will be removed. For
example, if you have more phenotyped participants than genotyped
participants in your study, <code><a href="../reference/create_design.html">plmmr::create_design()</a></code> will
create a matrix of data representing all the <em>genotyped</em> samples
that also have data in the supplied external phenotype file.</p></li>
<li><p>Create a design matrix that represents the nonsingular features
and the samples that have predictor and phenotype information available
(in the case where external data are supplied).</p></li>
<li><p>Standardize the design matrix so that all columns have mean of 0
and variance of 1.</p></li>
</ul>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tabitha K. Peter, Anna C. Reisetter, Patrick J. Breheny, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
